### PR TITLE
Proposal: Render refactor

### LINKED ISSUE #2: how to query __params field?
using `__params` to pass graphql parameters to a field resolver is nice, but what if I have an API which has `__params` field? How can I tell the typed-graphqlify that the `__params` is a field-not a params object?

I know it's not very likely, but it could happen.

### PR DESCRIPTION
**Pull request by luvies**
This is a proposal to refactor the old render system to a more explicit and flexible one, that supports extending very easily. Other than a removal of a comment, I haven't changed the tests, meaning that this is pretty much a non-breaking change, unless people were using values directly instead of using the `types` helper. The main features are:

- Rendering is now done through a `render` function that then calls other functions in a recursive manner to handle rendering different kinds of structures
- Inline fragments are now a pure object instead of a class to keep things consistent
- Scalars can now support parameters
  - While I haven't actually changed the API from an external perspective, the way scalars are handled internally now supports us putting metadata with them
  - This means a `params` or `$` function in future can interface with this and param rendering will work (this would mean a change to how object params are stored, however it would be via a similar system (mentioned below))
- We now attempt to render the minimum amount of text needed
  - The renderer will forgo adding spaces between things like brackets, since they don't need them

To support params on objects, we can simply add a property to them using the `paramsSymbol` Symbol, and then render the value of that (it's a simple change that I haven't done, since I didn't want to change the external API for this (I can make the change if requested to close #2 in this PR as well)). Part of the reason for doing it like this, instead of using the WeakMap system as suggested, is that it means that you can provide parameters for scalars using the same system, which wouldn't be possible otherwise (since you can't use primitives in WeakMaps).

### CODE REVIEW

**Comment by luvies** [comment]
wait this will also fix #41

File: src/render.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+ */
+export enum GraphQLType {
+  SCALAR,
+  INLINE_FRAGMENT,
+}
+
+/**
+ * The symbol to use to store the object GQL type
+ */
+export const typeSymbol = Symbol('GraphQL Type')
```
Comment (acro5piano) [inline src/render.ts]: Can you add newline here?

**Comment by acro5piano** [comment]
And I would like to solve #2 and release 2.1.0 (stable) so I really appreciate your help!

The roadmap is:

- Merge this PR while solving #41 
- Move on #2 and #38 
- Fix documents
- Release 2.1.0

Off topic Note: I also am wondering we can change the API simpler, like

```ts
import { query, mutation } from 'typed-graphqlify'
query('getUser', {
  // ...
})
```

Instead of 

```ts
import { graphqlify } from 'typed-graphqlify'
graphqlify.query('getUser', {
  // ...
})
```

However, further discussion is needed, so later we can decide it.

File: src/render.ts (hunk 1 above)
Comment (luvies) [reply src/render.ts]: sure thing

**Comment by luvies** [comment]
Over the next day or 2, I can look at closing #2 and #38 (since they're closely related to this PR anyway). I would say that we should probably adjust the imports, since it would support doing something like this still:

```ts
import * as graphqlify from 'typed-graphqlify'
graphqlify.query('getUser', {
  // ...
})
```

Meaning that depending code can be updated very easily. I'd open an issue for it for now, though

File: src/__test__/index.test.ts [hunk 2]
```diff
@@ -498,4 +491,67 @@ describe('graphqlify', () => {
...
+            const
+            oneOf
+            custom
+          }
+        }
+      }
+    `)
+  })
+
+  it('properly errors on invalid input', () => {
```
Comment (acro5piano) [inline src/__test__/index.test.ts]: Thanks!!

**Comment by acro5piano** [comment]
Thank you very much!

**Pull request merged**